src/Compilers/CSharp/Portable/Binder/WithUsingNamespacesAndTypesBinder.cs

```diff
@@ -0,0 +1,358 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// A binder that represents a scope introduced by 'using' namespace or type directives and deals with looking up names in it.
+    /// </summary>
+    internal abstract class WithUsingNamespacesAndTypesBinder : Binder
+    {
+        private readonly bool _withImportChainEntry;
+        private ImportChain? _lazyImportChain;
+
+        protected WithUsingNamespacesAndTypesBinder(Binder next, bool withImportChainEntry)
+            : base(next)
+        {
+            _withImportChainEntry = withImportChainEntry;
+        }
+
+        internal bool WithImportChainEntry => _withImportChainEntry;
```

Comment:
Is this property only used for an assertion in `WithExternAndUsingAliasesBinder.WithExternAndUsingAliasesBinder`? If so, consider `#if DEBUG`.